src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs

```diff
@@ -0,0 +1,256 @@
+            operationBlockStartAnalysisContext.RegisterOperationAction(
+                context => CollectPotentialDiagnosticOperations(
+                    context,
+                    wellKnownSymbolsInfo,
+                    potentialDiagnosticOperationsBuilder),
+                OperationKind.ParameterReference,
+                OperationKind.LocalReference);
+
+            operationBlockStartAnalysisContext.RegisterOperationBlockEndAction(
+                context => Analyze(
+                    context,
+                    wellKnownTypeProvider,
+                    wellKnownSymbolsInfo,
+                    potentialDiagnosticOperationsBuilder));
+        }
+
+        private static void CollectPotentialDiagnosticOperations(
+            OperationAnalysisContext context,
+            WellKnownSymbolsInfo wellKnownSymbolsInfo,
+            ImmutableHashSet<IOperation>.Builder builder)
+        {
+            var operation = context.Operation;
+            if (IsDeferredType(operation.Type.OriginalDefinition, wellKnownSymbolsInfo.AdditionalDeferredTypes))
+            {
+                var isEnumerated = IsOperationEnumeratedByMethodInvocation(operation, wellKnownSymbolsInfo) || IsOperationEnumeratedByForEachLoop(operation, wellKnownSymbolsInfo);
+                if (isEnumerated)
+                    builder.Add(operation);
+            }
+        }
+
+        private void Analyze(
+            OperationBlockAnalysisContext context,
+            WellKnownTypeProvider wellKnownTypeProvider,
+            WellKnownSymbolsInfo wellKnownSymbolsInfo,
+            ImmutableHashSet<IOperation>.Builder potentialDiagnosticOperationsBuilder)
+        {
+            var potentialDiagnosticOperations = potentialDiagnosticOperationsBuilder.ToImmutable();
+            if (potentialDiagnosticOperations.IsEmpty)
+            {
+                return;
+            }
+
+            var cfg = context.OperationBlocks.GetControlFlowGraph();
+            if (cfg == null)
+            {
+                return;
+            }
+
+            // In CFG blocks there is no foreach loop related Operation, so use the
+            // the GetEnumerator method to find the foreach loop
+            var getEnumeratorSymbol = wellKnownTypeProvider
+                .GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIEnumerable1)
+                ?.GetMembers(WellKnownMemberNames.GetEnumeratorMethodName).FirstOrDefault() as IMethodSymbol;
+
+            var analysisResult = GlobalFlowStateDictionaryAnalysis.TryGetOrComputeResult(
+                cfg,
+                context.OwningSymbol,
+                analysisContext => CreateOperationVisitor(
+                    analysisContext,
+                    wellKnownSymbolsInfo,
+                    getEnumeratorSymbol),
+                wellKnownTypeProvider,
+                context.Options,
+                MultipleEnumerableDescriptor,
+                pessimisticAnalysis: false);
```

Comment:
Add a comment as to why we want optimistic analysis here?

Reply:
Acutally after some thinking I feel it should be true here. Will add comments

Reply:
Added